File: Go-Solutions/nowcoder/jz24.go
package nowcoder

func FindPath(root *TreeNode, sum int) [][]int {
	// write code here
	if root == nil {
		return [][]int{}
	}
	res := make([][]int, 0)
	list := make([]int, 0)
	dfs(root, &res, &list, sum)
	return res
}

func dfs(node *TreeNode, res *[][]int, list *[]int, leftToSum int) {
	if node == nil {
		return
	}

	if node.Val > leftToSum {
		return
	}

	*list = append(*list, node.Val)
	if node.Val == leftToSum && node.Left == nil && node.Right == nil {
		listToAdd := make([]int, len(*list))
		for i := 0; i < len(*list); i++ {
			listToAdd[i] = (*list)[i]
		}
		*res = append(*res, listToAdd)
	} else {
		dfs(node.Left, res, list, leftToSum-node.Val)
		dfs(node.Right, res, list, leftToSum-node.Val)
	}
	*list = (*list)[:len(*list)-1]
}

File: Java-Kotlin/src/main/java/jianzhi/jz24.java
package jianzhi;

import common.datastructure.TreeNode;

import java.util.ArrayList;

public class jz24 {

    private static void dfs(TreeNode node, ArrayList<ArrayList<Integer>> res, ArrayList<Integer> list, int leftToTarget) {
        if (node == null) {
            return;
        }

        if (node.val > leftToTarget) {
            return;
        }

        list.add(node.val);
        if (node.val == leftToTarget && node.left == null && node.right == null) {
            res.add(new ArrayList<>(list));
        } else {
            dfs(node.left, res, list, leftToTarget - node.val);
            dfs(node.right, res, list, leftToTarget - node.val);
        }
        list.remove(list.size() - 1);
    }

    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root, int target) {
        if (root == null) {
            return new ArrayList<>();
        }

        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        ArrayList<Integer> list = new ArrayList<>();
        dfs(root, res, list, target);
        return res;
    }
}
